keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,1179 @@
+
+In beta, we intend to introduce support for greater compatibility version skew
+(specifically, N-3) so that it would be possible to skip binary versions while
+still performing a stepwise upgrade of Kubernetes control-plane. For example:
+
+- (starting point) binary-version 1.28 (compat-version 1.28)
+- upgrade binary-version to 1.31 (compat-version stays at 1.28 - this is our skip-level binary upgrade)
+- keep binary-version 1.31 while upgrading compat-version to 1.29 (stepwise upgrade of compatibility)
+- keep binary-version 1.31 while upgrading compat-version to 1.30 (stepwise upgrade of compatibility)
+- keep binary-version 1.31 while upgrading compat-version to 1.31 (stepwise upgrade of compatibility)
+
+Benefits to upgrading binary version independent of compatibility version:
+
+- A skip-version binary upgrade that transitions through compat versions has
+  higher probability of bugs in those intermediate binary versions already being
+  fixed.
+- During an upgrade, it is possible upgrade the new binary version while the
+  compatibility version is fixed (e.g. `binaryVersion: 1.28` -> `{binaryVersion: 1.29, compatVersion: 1.28}`).
+  This allow differences in the binary (bugs fixed or introduced, performance
+  changes, ...) to be introduced into a cluster and verified safe before
+  allowing access to the new APIs and features new version. Once the binary
+  version has been deamed safe, the compatibility version can then be upgraded.
+- Any upgrade system that successfully detects failures between upgrade steps
+  can report which upgrade step failed. This makes it easier to diagnose the
+  failures, because there are fewer possible causes of the failure. (There's a
+  huge difference between "A cluster upgrade failed" and "A cluster upgrade
+  failed when only the apiserver's binary version changed").
+- For upgrades where multiple failures occur, this increases the odds those
+  failures are split across steps. An upgrade system that is able to pause after
+  a step where failures are detected allow for failures at that step to be
+  addressed before proceeding to subsequent steps. These failures can be
+  addressed without the disruption and "noise" from failures in subsequent
+  steps.
+- A compatibility version rollback can be performed without changing binary
+  version.
+
+
+### Goals
+
+- Introduce the metadata necessary to configure features/APIs/storage-versions/CEL
+  features to match the behavior of an older Kubernetes release version
+- A Kubernetes binary with compatibility version set to N, will pass the
+  conformance and e2e tests from Kubernetes release version N.
+- A Kubernetes binary with compatibility version set to N does not enable any
+  changes (storage versions, CEL feature, features) that would prevent it
+  from being rolled back to N-1.
+- The most recent Kubernetes version supports compatibility version being set to
+  the full range of supported versions (N..N-3).
```

Comment:
Just clarifying that this means that for a given version `N` of k8s, that it would be an error to set `--compatibility-version` outside of the inclusive range `[N-3, N]`?

Reply:
Yes. It would be a invalid flag value type of error. The integration test grid below mentions this briefly too.